WireAPI/Sources/WireAPI/Network/PushChannel/PushChannel.swift

```diff
@@ -17,19 +17,141 @@
+    func open() async throws -> AsyncThrowingStream<UpdateEventEnvelope, Error> {
+        print("opening new push channel")
+        let webSocket = webSocketProvider.makeWebSocket(with: request)
+        self.webSocket = webSocket
+        let decoder = JSONDecoder()
+
+        return webSocket.makeStream().map { [weak self] message in
+            do {
+                switch message {
+                case .data(let data):
+                    print("received web socket data, decoding...")
+                    let envelope = try decoder.decode(UpdateEventEnvelopeV0.self, from: data)
+                    return envelope.toAPIModel()
+
+                case .string:
+                    print("received web socket string, ignoring...")
+                    throw PushChannelError.receivedInvalidMessage
+
+                @unknown default:
+                    print("received web socket message, ignoring...")
+                    throw PushChannelError.receivedInvalidMessage
+                }
+            } catch {
+                print("failed to get next web socket message: \(error)")
+                self?.close()
+                throw error
+            }
+        }.toStream()
+    }
+
+    func close() {
+        guard let webSocket else { return }
+        print("closing push channel")
+        webSocket.close()
+        self.webSocket = nil
+    }
+
+}
+
+extension PushChannel: URLSessionWebSocketDelegate {
+
+    func urlSession(
+        _ session: URLSession,
+        webSocketTask: URLSessionWebSocketTask,
+        didOpenWithProtocol protocol: String?
+    ) {
+        print("web socket task did open")
+    }
+
+    func urlSession(
+        _ session: URLSession,
+        webSocketTask: URLSessionWebSocketTask,
+        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
+        reason: Data?
+    ) {
+        close()
+    }
+
+}
+
+extension PushChannel: URLSessionDataDelegate {
+
+    func urlSession(
+        _ session: URLSession,
+        task: URLSessionTask,
+        didCompleteWithError error: Error?
+    ) {
+        if let error {
+            print("web socket task did complete with error: \(error)")
+        } else {
+            print("web socket task did complete")
+        }
+
+        close()
+    }
+
+    func urlSession(
+        _ session: URLSession,
+        task: URLSessionTask,
+        didReceive challenge: URLAuthenticationChallenge,
+        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
+    ) {
+        print("web socket task did receive challenge")
+
+        let protectionSpace = challenge.protectionSpace
+
+        guard protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust else {
+            completionHandler(.performDefaultHandling, challenge.proposedCredential)
+            return
+        }
+
+        guard
+            protectionSpace.serverTrust != nil,
+            true // TODO: [WPB-10450] support certificate pinning
```

Comment:
Will we fix this before actually using this in the app?

Reply:
Yes but not in this PR.